<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>2 Getting started | Introduction to single-cell RNA-seq data analysis</title>
  <meta name="description" content="2 Getting started | Introduction to single-cell RNA-seq data analysis" />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="2 Getting started | Introduction to single-cell RNA-seq data analysis" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="2 Getting started | Introduction to single-cell RNA-seq data analysis" />
  
  
  

<meta name="author" content="Eddie Cano-Gamez" />


<meta name="date" content="2023-05-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="introduction.html"/>
<link rel="next" href="practical-studying-cd4-t-cells-from-human-blood.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#aims-of-this-course"><i class="fa fa-check"></i><b>1.1</b> Aims of this course</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#course-pre-requisites"><i class="fa fa-check"></i><b>1.2</b> Course pre-requisites</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#the-example-data-set"><i class="fa fa-check"></i><b>1.3</b> The example data set</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="getting-started.html"><a href="getting-started.html"><i class="fa fa-check"></i><b>2</b> Getting started</a>
<ul>
<li class="chapter" data-level="2.1" data-path="getting-started.html"><a href="getting-started.html#downloading-course-materials"><i class="fa fa-check"></i><b>2.1</b> Downloading course materials</a></li>
<li class="chapter" data-level="2.2" data-path="getting-started.html"><a href="getting-started.html#setting-up-the-correct-r-environment"><i class="fa fa-check"></i><b>2.2</b> Setting up the correct R environment</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="getting-started.html"><a href="getting-started.html#option-1-installing-all-libraries-automatically-with-renv"><i class="fa fa-check"></i><b>2.2.1</b> Option 1: installing all libraries automatically with renv</a></li>
<li class="chapter" data-level="2.2.2" data-path="getting-started.html"><a href="getting-started.html#option-2-installing-each-library-manually"><i class="fa fa-check"></i><b>2.2.2</b> Option 2: installing each library manually</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="getting-started.html"><a href="getting-started.html#verifying-set-up"><i class="fa fa-check"></i><b>2.3</b> Verifying set up</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html"><i class="fa fa-check"></i><b>3</b> Practical: Studying CD4+ T cells from human blood</a>
<ul>
<li class="chapter" data-level="3.1" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#overview"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#loading-data-and-r-libraries"><i class="fa fa-check"></i><b>3.2</b> Loading data and R libraries</a></li>
<li class="chapter" data-level="3.3" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#formatting-data-according-for-analysis"><i class="fa fa-check"></i><b>3.3</b> Formatting data according for analysis</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#creating-a-seurat-object"><i class="fa fa-check"></i><b>3.3.1</b> Creating a Seurat object</a></li>
<li class="chapter" data-level="3.3.2" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#how-are-single-cell-measurements-actually-handled-by-the-computer"><i class="fa fa-check"></i><b>3.3.2</b> How are single-cell measurements actually handled by the computer?</a></li>
<li class="chapter" data-level="3.3.3" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#adding-cell-annotations-to-our-matrix"><i class="fa fa-check"></i><b>3.3.3</b> Adding cell annotations to our matrix</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#removing-low-quality-cells-and-genes"><i class="fa fa-check"></i><b>3.4</b> Removing low quality cells and genes</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cell-filtering"><i class="fa fa-check"></i><b>3.4.1</b> Cell filtering</a></li>
<li class="chapter" data-level="3.4.2" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#gene-filtering"><i class="fa fa-check"></i><b>3.4.2</b> Gene filtering</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#analysing-gene-expression-data"><i class="fa fa-check"></i><b>3.5</b> Analysing gene expression data</a></li>
<li class="chapter" data-level="3.6" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#data-visualisation"><i class="fa fa-check"></i><b>3.6</b> Data visualisation</a>
<ul>
<li class="chapter" data-level="3.6.1" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>3.6.1</b> Principal component analysis (PCA)</a></li>
<li class="chapter" data-level="3.6.2" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#component-embedding-with-umap"><i class="fa fa-check"></i><b>3.6.2</b> Component embedding with UMAP</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#clustering-cells-into-groups"><i class="fa fa-check"></i><b>3.7</b> Clustering cells into groups</a></li>
<li class="chapter" data-level="3.8" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#identifying-cell-type-markers"><i class="fa fa-check"></i><b>3.8</b> Identifying cell type markers</a></li>
<li class="chapter" data-level="3.9" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#conclusion"><i class="fa fa-check"></i><b>3.9</b> Conclusion</a></li>
<li class="chapter" data-level="3.10" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#r-session"><i class="fa fa-check"></i><b>3.10</b> R session</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to single-cell RNA-seq data analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="getting-started" class="section level1 hasAnchor" number="2">
<h1><span class="header-section-number">2</span> Getting started<a href="getting-started.html#getting-started" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Before we get going with data analysis, we will need to make sure that your computing environment is correctly set up and that you have all of the necessary requirements for this course to run smoothly.</p>
<p>Let’s review these one by one.</p>
<div id="downloading-course-materials" class="section level2 hasAnchor" number="2.1">
<h2><span class="header-section-number">2.1</span> Downloading course materials<a href="getting-started.html#downloading-course-materials" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We will begin by downloading the example data set used throughout this course.</p>
<p>To do so, open your terminal window (i.e. the command line) and run the following command:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="getting-started.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/eddiecg/single-cell-course_2023.git</span></code></pre></div>
<p>This will clone the GitHub repostiory associated with this course in your local computer, thus automatically downloading all necessary data sets.</p>
<p>To make sure all files downloaded successfully, move into the newly created directory by running the following command in the terminal.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="getting-started.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> single-cell-course_2023</span></code></pre></div>
<p>Finally, run the following in your terminal:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="getting-started.html#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-l</span> data</span></code></pre></div>
<pre><code>## total 38120
## -rw-r--r--@ 1 eddiecano  staff     29797 May 30 13:36 barcodes.tsv.gz
## -rw-r--r--@ 1 eddiecano  staff    262330 May 30 13:37 cell-annotations.tsv.gz
## -rw-r--r--@ 1 eddiecano  staff     81256 May 30 13:36 features.tsv.gz
## -rw-r--r--@ 1 eddiecano  staff  18524082 May 30 13:34 matrix.mtx.gz</code></pre>
<p>The output from this command should be a lsit of all the files required for this course. These include:</p>
<p><strong>matrix.mtx.gz</strong> A single-cell gene expression matrix containing measurements for 20,953 genes across 5,269 cells.</p>
<p><strong>barcodes.tsv.gz</strong> The cell IDs corresponding to each column of the expression matrix.</p>
<p><strong>features.tsv.gz</strong> The gene names corresponding to each row of the expression matrix.</p>
<p><strong>cell-annotations.tsv.gz</strong> A table with annotations and metadata associated with each single cell in the expression matrix.</p>
</div>
<div id="setting-up-the-correct-r-environment" class="section level2 hasAnchor" number="2.2">
<h2><span class="header-section-number">2.2</span> Setting up the correct R environment<a href="getting-started.html#setting-up-the-correct-r-environment" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Now that you have downloaded the required data, we can proceed to create an R project specifically for this course and to install the R libraries required for our analysis.</p>
<p>Start by opening RStudio and creating an R project. You can do this by selecting:</p>
<p>File &gt; New Project</p>
<p>Make sure to create the project within the folder you just downloaded (i.e. single-cell-course_2023)</p>
<p>Once your project has been created, you can install the required R libraries. You can do this by following one of two options:</p>
<ol style="list-style-type: decimal">
<li><p>Using the renv package to automatically install all libraries (recommended): By following these instructions you will be installing the exact same software and library versions that I used to create this course, thus minimising any potential technical issues and differences.</p></li>
<li><p>Manually installing each library: This is easier to do, but may result in your library versions being slightly different to those I used to create the course.</p></li>
</ol>
<div id="option-1-installing-all-libraries-automatically-with-renv" class="section level3 hasAnchor" number="2.2.1">
<h3><span class="header-section-number">2.2.1</span> Option 1: installing all libraries automatically with renv<a href="getting-started.html#option-1-installing-all-libraries-automatically-with-renv" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To iuse this option, you will first need to make sure renv is properly installed. You can do so by running the following lines:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="getting-started.html#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;remotes&quot;</span>))</span>
<span id="cb5-2"><a href="getting-started.html#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;remotes&quot;</span>)</span>
<span id="cb5-3"><a href="getting-started.html#cb5-3" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;rstudio/renv&quot;</span>)</span></code></pre></div>
<p>Once renv has installed, you can then initialise your R project and automatically install all the rquired libraries, which will by default match the versions I used. You can do so by running the following line:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="getting-started.html#cb6-1" aria-hidden="true" tabindex="-1"></a>renv<span class="sc">::</span><span class="fu">init</span>()</span>
<span id="cb6-2"><a href="getting-started.html#cb6-2" aria-hidden="true" tabindex="-1"></a>renv<span class="sc">::</span><span class="fu">restore</span>()</span></code></pre></div>
<p>NOTE: if asked whether you’d like to restore the project from the existing lockfile, select yes (e.g. 1: Restore the project from the lockfile)</p>
</div>
<div id="option-2-installing-each-library-manually" class="section level3 hasAnchor" number="2.2.2">
<h3><span class="header-section-number">2.2.2</span> Option 2: installing each library manually<a href="getting-started.html#option-2-installing-each-library-manually" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Alternatively, if the above method does not work for you, you can also install each library manually using base R. to do so, simply run the following lines:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="getting-started.html#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;tidyverse&quot;</span>)</span>
<span id="cb7-2"><a href="getting-started.html#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;Seurat&quot;</span>)</span>
<span id="cb7-3"><a href="getting-started.html#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;hexbin&quot;</span>)</span>
<span id="cb7-4"><a href="getting-started.html#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;Matrix&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="verifying-set-up" class="section level2 hasAnchor" number="2.3">
<h2><span class="header-section-number">2.3</span> Verifying set up<a href="getting-started.html#verifying-set-up" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>To verify that all libraries were indeed correctly installed, try loading them:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="getting-started.html#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;tidyverse&quot;</span>)</span>
<span id="cb8-2"><a href="getting-started.html#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;Seurat&quot;</span>)</span>
<span id="cb8-3"><a href="getting-started.html#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;hexbin&quot;</span>)</span>
<span id="cb8-4"><a href="getting-started.html#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;Matrix&quot;</span>)</span></code></pre></div>
<p>If the code block above ran successfully and without any errors, then you are ready to start this course!</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="practical-studying-cd4-t-cells-from-human-blood.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
