<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Introduction to single-cell RNA-seq data analysis</title>
  <meta name="description" content="Introduction to single-cell RNA-seq data analysis" />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="Introduction to single-cell RNA-seq data analysis" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Introduction to single-cell RNA-seq data analysis" />
  
  
  

<meta name="author" content="Eddie Cano-Gamez" />


<meta name="date" content="2023-05-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="getting-started.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#aims-of-this-course"><i class="fa fa-check"></i><b>1.1</b> Aims of this course</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#course-pre-requisites"><i class="fa fa-check"></i><b>1.2</b> Course pre-requisites</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#the-example-data-set"><i class="fa fa-check"></i><b>1.3</b> The example data set</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="getting-started.html"><a href="getting-started.html"><i class="fa fa-check"></i><b>2</b> Getting started</a>
<ul>
<li class="chapter" data-level="2.1" data-path="getting-started.html"><a href="getting-started.html#downloading-course-materials"><i class="fa fa-check"></i><b>2.1</b> Downloading course materials</a></li>
<li class="chapter" data-level="2.2" data-path="getting-started.html"><a href="getting-started.html#setting-up-the-correct-r-environment"><i class="fa fa-check"></i><b>2.2</b> Setting up the correct R environment</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="getting-started.html"><a href="getting-started.html#option-1-installing-all-libraries-automatically-with-renv"><i class="fa fa-check"></i><b>2.2.1</b> Option 1: installing all libraries automatically with renv</a></li>
<li class="chapter" data-level="2.2.2" data-path="getting-started.html"><a href="getting-started.html#option-2-installing-each-library-manually"><i class="fa fa-check"></i><b>2.2.2</b> Option 2: installing each library manually</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="getting-started.html"><a href="getting-started.html#verifying-set-up"><i class="fa fa-check"></i><b>2.3</b> Verifying set up</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html"><i class="fa fa-check"></i><b>3</b> Practical: Studying CD4+ T cells from human blood</a>
<ul>
<li class="chapter" data-level="3.1" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#overview"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#loading-data-and-r-libraries"><i class="fa fa-check"></i><b>3.2</b> Loading data and R libraries</a></li>
<li class="chapter" data-level="3.3" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#formatting-data-for-analysis"><i class="fa fa-check"></i><b>3.3</b> Formatting data for analysis</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#creating-a-seurat-object"><i class="fa fa-check"></i><b>3.3.1</b> Creating a Seurat object</a></li>
<li class="chapter" data-level="3.3.2" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#how-are-single-cell-measurements-actually-handled-by-the-computer"><i class="fa fa-check"></i><b>3.3.2</b> How are single-cell measurements actually handled by the computer?</a></li>
<li class="chapter" data-level="3.3.3" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#adding-cell-annotations-to-our-matrix"><i class="fa fa-check"></i><b>3.3.3</b> Adding cell annotations to our matrix</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#removing-low-quality-cells-and-genes"><i class="fa fa-check"></i><b>3.4</b> Removing low quality cells and genes</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cell-filtering"><i class="fa fa-check"></i><b>3.4.1</b> Cell filtering</a></li>
<li class="chapter" data-level="3.4.2" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#gene-filtering"><i class="fa fa-check"></i><b>3.4.2</b> Gene filtering</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#analysing-gene-expression-data"><i class="fa fa-check"></i><b>3.5</b> Analysing gene expression data</a></li>
<li class="chapter" data-level="3.6" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#data-visualisation"><i class="fa fa-check"></i><b>3.6</b> Data visualisation</a>
<ul>
<li class="chapter" data-level="3.6.1" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>3.6.1</b> Principal component analysis (PCA)</a></li>
<li class="chapter" data-level="3.6.2" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#component-embedding-with-umap"><i class="fa fa-check"></i><b>3.6.2</b> Component embedding with UMAP</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#clustering-cells-into-groups"><i class="fa fa-check"></i><b>3.7</b> Clustering cells into groups</a></li>
<li class="chapter" data-level="3.8" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#identifying-cell-type-markers"><i class="fa fa-check"></i><b>3.8</b> Identifying cell type markers</a></li>
<li class="chapter" data-level="3.9" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#conclusion"><i class="fa fa-check"></i><b>3.9</b> Conclusion</a></li>
<li class="chapter" data-level="3.10" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#r-session"><i class="fa fa-check"></i><b>3.10</b> R session</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to single-cell RNA-seq data analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">Introduction to single-cell RNA-seq data analysis</h1>
<p class="author"><em>Eddie Cano-Gamez</em></p>
<p class="date"><em>2023-05-30</em></p>
</div>
<div id="introduction" class="section level1 hasAnchor" number="1">
<h1><span class="header-section-number">1</span> Introduction<a href="introduction.html#introduction" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Wellcome to the <strong>Introduction to Single-Cell RNA-Sequencing Data Analysis course</strong>. My name is Eddie Cano-Gamez, and I am a researcher in immunology and single-cell biology. In this course I aim to help you getting started with analysing single-cell RNA-sequencing (scRNA-seq) data in R.</p>
<div id="aims-of-this-course" class="section level2 hasAnchor" number="1.1">
<h2><span class="header-section-number">1.1</span> Aims of this course<a href="introduction.html#aims-of-this-course" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this cousre we will cover basic concepts such as:</p>
<ul>
<li>Which types of files single-cell RNA-seq data are stored in</li>
<li>How to load scRNA-seq data into R for analysis</li>
<li>How to create suitable R objects to make scRNA-seq data analysis as easy as possbile</li>
<li>The basic quality checks which you should perform in any scRNA-seq data set</li>
<li>How to visualise cells based on their gene expression profiles</li>
</ul>
<p>We will also briefly touch on a few more advanced concepts like:
- Linear and non-linear dimensionality reduction techniques
- Unsupervised clustering of cells into cell clusters
- Subsequent annotation of cell clusters into cell types based on expression of marker genes</p>
<p>Unfortnuately, due to time constrains we will not be able to cover advanced topics, such as:
- Trajectory analysis and pseudotime
- Multi-omics analysis
- Batch integration</p>
<p>However, you can always reach me at my institutional email if you have any questions regarding these topics: <a href="mailto:ecg@well.ox.ac.uk" class="email">ecg@well.ox.ac.uk</a></p>
</div>
<div id="course-pre-requisites" class="section level2 hasAnchor" number="1.2">
<h2><span class="header-section-number">1.2</span> Course pre-requisites<a href="introduction.html#course-pre-requisites" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>You <strong>do not</strong> need to be fluent in R or an expert programming to understand this course. However, some familiarity with R would greatly increase your chances of following the material and learning something new from it. Thus, if you have never programmed in R before, I highly recommend you look at the basic R programming material available in my GitHub before you attend this course:</p>
<p><a href="https://eddiecg.github.io/Data-analysis-in-R/introduction-to-programming-in-r.html" class="uri">https://eddiecg.github.io/Data-analysis-in-R/introduction-to-programming-in-r.html</a></p>
<p>To follow this course, you will need to install the following:</p>
<ul>
<li>R (version 4 or later): see <a href="https://www.r-project.org/" class="uri">https://www.r-project.org/</a></li>
<li>RStudio: see <a href="https://posit.co/download/rstudio-desktop/" class="uri">https://posit.co/download/rstudio-desktop/</a></li>
<li>The R libraries required for this course (see <em>Section 2: Getting Started</em>)</li>
</ul>
<p>NOTE: It’s important you do this <strong>before</strong> the course, so that you don’t use up other participant’s time in sorting out software installation issues.</p>
</div>
<div id="the-example-data-set" class="section level2 hasAnchor" number="1.3">
<h2><span class="header-section-number">1.3</span> The example data set<a href="introduction.html#the-example-data-set" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The practical exercises presented in this course are based on a medium-sized data set of circulating CD4+ T cells obtained from healthy volunteers. This data was generated at the Wellcome Sanger Institute using the 10X scRNA-seq platform and is publicly available.</p>
<p>While we won’t be using the full data set (which contains close to 40,000 cells and would be unfeasible to work with in a local laptop), this study has been previously published in Nature Communications (<a href="https://doi.org/10.1038/s41467-020-15543-y" class="uri">https://doi.org/10.1038/s41467-020-15543-y</a>), and you can access the full data set in the resources associated with that publication.</p>
<p>I’ve made this example data set available via my GitHub account, and the following vignettes will show you how to download it.</p>
<p>Let’s get going!</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="getting-started.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
