<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3 Practical: Studying CD4+ T cells from human blood | Introduction to single-cell RNA-seq data analysis</title>
  <meta name="description" content="3 Practical: Studying CD4+ T cells from human blood | Introduction to single-cell RNA-seq data analysis" />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="3 Practical: Studying CD4+ T cells from human blood | Introduction to single-cell RNA-seq data analysis" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3 Practical: Studying CD4+ T cells from human blood | Introduction to single-cell RNA-seq data analysis" />
  
  
  

<meta name="author" content="Eddie Cano-Gamez" />


<meta name="date" content="2023-05-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="getting-started.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#aims-of-this-course"><i class="fa fa-check"></i><b>1.1</b> Aims of this course</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#course-pre-requisites"><i class="fa fa-check"></i><b>1.2</b> Course pre-requisites</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#the-example-data-set"><i class="fa fa-check"></i><b>1.3</b> The example data set</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="getting-started.html"><a href="getting-started.html"><i class="fa fa-check"></i><b>2</b> Getting started</a>
<ul>
<li class="chapter" data-level="2.1" data-path="getting-started.html"><a href="getting-started.html#downloading-course-materials"><i class="fa fa-check"></i><b>2.1</b> Downloading course materials</a></li>
<li class="chapter" data-level="2.2" data-path="getting-started.html"><a href="getting-started.html#setting-up-the-correct-r-environment"><i class="fa fa-check"></i><b>2.2</b> Setting up the correct R environment</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="getting-started.html"><a href="getting-started.html#option-1-installing-all-libraries-automatically-with-renv"><i class="fa fa-check"></i><b>2.2.1</b> Option 1: installing all libraries automatically with renv</a></li>
<li class="chapter" data-level="2.2.2" data-path="getting-started.html"><a href="getting-started.html#option-2-installing-each-library-manually"><i class="fa fa-check"></i><b>2.2.2</b> Option 2: installing each library manually</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="getting-started.html"><a href="getting-started.html#verifying-set-up"><i class="fa fa-check"></i><b>2.3</b> Verifying set up</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html"><i class="fa fa-check"></i><b>3</b> Practical: Studying CD4+ T cells from human blood</a>
<ul>
<li class="chapter" data-level="3.1" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#overview"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#loading-data-and-r-libraries"><i class="fa fa-check"></i><b>3.2</b> Loading data and R libraries</a></li>
<li class="chapter" data-level="3.3" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#formatting-data-according-for-analysis"><i class="fa fa-check"></i><b>3.3</b> Formatting data according for analysis</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#creating-a-seurat-object"><i class="fa fa-check"></i><b>3.3.1</b> Creating a Seurat object</a></li>
<li class="chapter" data-level="3.3.2" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#how-are-single-cell-measurements-actually-handled-by-the-computer"><i class="fa fa-check"></i><b>3.3.2</b> How are single-cell measurements actually handled by the computer?</a></li>
<li class="chapter" data-level="3.3.3" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#adding-cell-annotations-to-our-matrix"><i class="fa fa-check"></i><b>3.3.3</b> Adding cell annotations to our matrix</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#removing-low-quality-cells-and-genes"><i class="fa fa-check"></i><b>3.4</b> Removing low quality cells and genes</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cell-filtering"><i class="fa fa-check"></i><b>3.4.1</b> Cell filtering</a></li>
<li class="chapter" data-level="3.4.2" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#gene-filtering"><i class="fa fa-check"></i><b>3.4.2</b> Gene filtering</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#analysing-gene-expression-data"><i class="fa fa-check"></i><b>3.5</b> Analysing gene expression data</a></li>
<li class="chapter" data-level="3.6" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#data-visualisation"><i class="fa fa-check"></i><b>3.6</b> Data visualisation</a>
<ul>
<li class="chapter" data-level="3.6.1" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>3.6.1</b> Principal component analysis (PCA)</a></li>
<li class="chapter" data-level="3.6.2" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#component-embedding-with-umap"><i class="fa fa-check"></i><b>3.6.2</b> Component embedding with UMAP</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#clustering-cells-into-groups"><i class="fa fa-check"></i><b>3.7</b> Clustering cells into groups</a></li>
<li class="chapter" data-level="3.8" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#identifying-cell-type-markers"><i class="fa fa-check"></i><b>3.8</b> Identifying cell type markers</a></li>
<li class="chapter" data-level="3.9" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#conclusion"><i class="fa fa-check"></i><b>3.9</b> Conclusion</a></li>
<li class="chapter" data-level="3.10" data-path="practical-studying-cd4-t-cells-from-human-blood.html"><a href="practical-studying-cd4-t-cells-from-human-blood.html#r-session"><i class="fa fa-check"></i><b>3.10</b> R session</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to single-cell RNA-seq data analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="practical-studying-cd4-t-cells-from-human-blood" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">3</span> Practical: Studying CD4+ T cells from human blood<a href="practical-studying-cd4-t-cells-from-human-blood.html#practical-studying-cd4-t-cells-from-human-blood" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Now that you have downloaded the necessary data and installed any pre-required libraries, we can begin our analysis.</p>
<div id="overview" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> Overview<a href="practical-studying-cd4-t-cells-from-human-blood.html#overview" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this course, we will analyse a data set of CD4+ T cells in circulating blood obtained from four healthy volunteers. This data comes from a study published in 2020 by myself and colleagues, and titled <strong>Single-cell transcriptomics identifies an effectorness gradient shaping the response of CD4+ T cells to cytokines</strong>.</p>
<p>For this study, we drew blood from healthy research participants and used a combination of laboratory techniques to isolate naive and memory CD4+ T cells from these blood samples. CD4+ T cells are one of the main participants in the immune response: they protect us against pathogens by telling the rest of the immune system what to do. However, we do not yet fully understand just how diverse these cells are and how much they can adapt to the environment in which they are present. That’s what our study tried to find out.</p>
<p>The experimental design of our study is summarised in the following diagram:</p>
<div class="figure">
<img src="figures/experimental-design.png" alt="" />
<p class="caption">Experimental design: we isolated naive and memory CD4 possitive T cells from four healthy volunteers and profiled them at several points in time</p>
</div>
<p>While our study sequenced more than 40,000 cells, that scale of data would be difficult to deal with in a regular laptop without taking up most of your computer memory. Thus, I have prepared a smaller subset of that data set, which will limit our analysis for this course to only one of the conditions we studied: cells obtained directly from blood and sequenced without any further treatment. This is what’s contained in the example data set you downloaded from GitHub.</p>
</div>
<div id="loading-data-and-r-libraries" class="section level2 hasAnchor" number="3.2">
<h2><span class="header-section-number">3.2</span> Loading data and R libraries<a href="practical-studying-cd4-t-cells-from-human-blood.html#loading-data-and-r-libraries" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>To begin with, let’s load the following R libraries:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb9-2"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb9-3"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(hexbin)</span></code></pre></div>
<p>Next, let’s load the data you downloaded from GitHub into R. We do this by using the <em>Read10X</em> function in the Seurat library. Note how this function goes to the loaction you are pointing it to and finds three files: the expression table itself (matrix.mtx), the gene names (features.tsv), and the cell names (barcodes.tsv). Next, it reads all three files into RAM and combines them together.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb10-1" aria-hidden="true" tabindex="-1"></a>t_cell_data <span class="ot">&lt;-</span> pbmc.data <span class="ot">&lt;-</span> <span class="fu">Read10X</span>(<span class="at">data.dir =</span> <span class="st">&quot;./data/&quot;</span>, <span class="at">gene.column=</span><span class="dv">1</span>)</span></code></pre></div>
<p>We will also read into memory the <em>cell-annotations.tsv</em> file, which contains information for each cell, based on the observations we published in our study.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb11-1" aria-hidden="true" tabindex="-1"></a>cell_anns <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> <span class="st">&quot;./data/cell-annotations.tsv.gz&quot;</span>, <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span></code></pre></div>
</div>
<div id="formatting-data-according-for-analysis" class="section level2 hasAnchor" number="3.3">
<h2><span class="header-section-number">3.3</span> Formatting data according for analysis<a href="practical-studying-cd4-t-cells-from-human-blood.html#formatting-data-according-for-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="creating-a-seurat-object" class="section level3 hasAnchor" number="3.3.1">
<h3><span class="header-section-number">3.3.1</span> Creating a Seurat object<a href="practical-studying-cd4-t-cells-from-human-blood.html#creating-a-seurat-object" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>While our data has been loaded into R successfully, it is still not in the most convenient format for us to analyse. Thus, we will use it to create a so called “Seurat Object”. This object is a data structure specifically designed for single-cell data analysis, and which will make our lives easier down the line. You can create a Seurat Object by running the following line:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb12-1" aria-hidden="true" tabindex="-1"></a>t_cells <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(</span>
<span id="cb12-2"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">counts =</span> t_cell_data, </span>
<span id="cb12-3"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">project =</span> <span class="st">&quot;CD4_T_cells&quot;</span>, </span>
<span id="cb12-4"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.cells =</span> <span class="dv">3</span>,</span>
<span id="cb12-5"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.features =</span> <span class="dv">200</span></span>
<span id="cb12-6"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb12-6" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>The resulting object looks as follows:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb13-1" aria-hidden="true" tabindex="-1"></a>t_cells</span></code></pre></div>
<pre><code>## An object of class Seurat 
## 14835 features across 5269 samples within 1 assay 
## Active assay: RNA (14835 features, 0 variable features)</code></pre>
<p>Note how R is telling us that our Seurat object contains information for 14,835 features (i.e. genes) and 5,269 samples (i.e. cells). This means that this little object contains more than 78 million numbers!</p>
<p>This is not untypical of single-cell experiments, which are one of the technologies capable of generating the most amount of data these days. So how is it possible to work with more than 78 million numbers without making our computer crash completely?</p>
</div>
<div id="how-are-single-cell-measurements-actually-handled-by-the-computer" class="section level3 hasAnchor" number="3.3.2">
<h3><span class="header-section-number">3.3.2</span> How are single-cell measurements actually handled by the computer?<a href="practical-studying-cd4-t-cells-from-human-blood.html#how-are-single-cell-measurements-actually-handled-by-the-computer" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Partly, this is due to a very clever computational trick. Let’s look at the first 10 rows and the first 10 columns of our data set:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb15-1" aria-hidden="true" tabindex="-1"></a>pbmc.data[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>## 10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
##                                  
## RP11-34P13.7  . . . . . . . . . .
## FO538757.2    1 . . . . 1 . . 1 1
## AP006222.2    . . . . . . . . . .
## RP4-669L17.10 . . . . . . . . . .
## RP11-206L10.9 . . . . . . . . . .
## LINC00115     . . . . . . . . . .
## FAM41C        . . . . . . . . . .
## NOC2L         . . . . . . 1 . . 1
## KLHL17        . . . . . . . . . .
## PLEKHN1       . . . . . . . . . .</code></pre>
<p>Let’s take some time to understand what is going on here. This is an expression matrix. Each column represents one cell, and each row represents one gene. The numbers in the matrixtell us how many molecules (which is proportional to the number of mRNAs) of each gene were detected in each cell.</p>
<p>You’ll notice a large number of dots (.). These dots represent zeros. The majority of the data generated by a single-cell experiment is made up of zeros: genes that were either not expressed in that cell or could not be detected due to them being present at a very low amount. We call this phenomenon (too many zeros) <em>sparsity</em>.</p>
<p>In order to save memory and make our analysis run much faster, Seurat does not store zeros in memory at all: they simply are left empty with a placeholder (represented here by the dot).</p>
<p>If we wanted to, we could transform these dots into actual zeros, with the following line:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.matrix</span>(pbmc.data[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span></code></pre></div>
<pre><code>##               N_resting_AAACCTGAGCTGTCTA N_resting_AAACCTGTCACCACCT N_resting_AAACCTGTCCGTTGTC N_resting_AAACGGGAGGGTTCCC N_resting_AAACGGGCAACAACCT
## RP11-34P13.7                           0                          0                          0                          0                          0
## FO538757.2                             1                          0                          0                          0                          0
## AP006222.2                             0                          0                          0                          0                          0
## RP4-669L17.10                          0                          0                          0                          0                          0
## RP11-206L10.9                          0                          0                          0                          0                          0
## LINC00115                              0                          0                          0                          0                          0
## FAM41C                                 0                          0                          0                          0                          0
## NOC2L                                  0                          0                          0                          0                          0
## KLHL17                                 0                          0                          0                          0                          0
## PLEKHN1                                0                          0                          0                          0                          0
##               N_resting_AAACGGGCACAACGCC N_resting_AAACGGGGTGCCTGTG N_resting_AAAGATGCACGGCCAT N_resting_AAAGATGCATCGATTG N_resting_AAAGCAAAGCGTGAAC
## RP11-34P13.7                           0                          0                          0                          0                          0
## FO538757.2                             1                          0                          0                          1                          1
## AP006222.2                             0                          0                          0                          0                          0
## RP4-669L17.10                          0                          0                          0                          0                          0
## RP11-206L10.9                          0                          0                          0                          0                          0
## LINC00115                              0                          0                          0                          0                          0
## FAM41C                                 0                          0                          0                          0                          0
## NOC2L                                  0                          1                          0                          0                          1
## KLHL17                                 0                          0                          0                          0                          0
## PLEKHN1                                0                          0                          0                          0                          0</code></pre>
<p>However, if we measure the amount of memory taken up by the matrix when zeros are not stored versus when they are, we will notice the following:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">format</span>(<span class="fu">object.size</span>(pbmc.data), <span class="at">units=</span><span class="st">&quot;Mb&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;70.8 Mb&quot;</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">format</span>(<span class="fu">object.size</span>(<span class="fu">as.matrix</span>(pbmc.data)), <span class="at">units=</span><span class="st">&quot;Mb&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;844.1 Mb&quot;</code></pre>
<p>Storing zeros as numbers would make our matrix 12 times bigger! That is why they are left empty instead. It’s just more convenient.</p>
</div>
<div id="adding-cell-annotations-to-our-matrix" class="section level3 hasAnchor" number="3.3.3">
<h3><span class="header-section-number">3.3.3</span> Adding cell annotations to our matrix<a href="practical-studying-cd4-t-cells-from-human-blood.html#adding-cell-annotations-to-our-matrix" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Let’s now move on to the cell annotations table. Printing the first few lines of this table reveals the type of information stored in it:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cell_anns)</span></code></pre></div>
<pre><code>##                            cell.type cytokine.condition donor.id batch.10X nGene nUMI percent.mito     S.Score  G2M.Score Phase   cluster.id
## N_resting_AAACCTGAGCTGTCTA     Naive                UNS       D4         2  1163 4172   0.02349556 -0.13419873 -0.1592109    G1 TN (resting)
## N_resting_AAACCTGTCACCACCT     Naive                UNS       D4         2  1037 3690   0.02086721 -0.10175611 -0.2037066    G1 TN (resting)
## N_resting_AAACCTGTCCGTTGTC     Naive                UNS       D2         2  1245 4446   0.02790279 -0.14513081 -0.1642104    G1 TN (resting)
## N_resting_AAACGGGAGGGTTCCC     Naive                UNS       D4         2  1016 3913   0.01150895 -0.06949151 -0.1908101    G1 TN (resting)
## N_resting_AAACGGGCAACAACCT     Naive                UNS       D1         2  1005 3557   0.03964015 -0.12400660 -0.1433789    G1 TN (resting)
## N_resting_AAACGGGCACAACGCC     Naive                UNS       D2         2  1404 5758   0.03179291 -0.18880479 -0.1985836    G1 TN (resting)
##                            effectorness
## N_resting_AAACCTGAGCTGTCTA   0.15181238
## N_resting_AAACCTGTCACCACCT   0.03176291
## N_resting_AAACCTGTCCGTTGTC   0.11389687
## N_resting_AAACGGGAGGGTTCCC   0.34123968
## N_resting_AAACGGGCAACAACCT   0.01974084
## N_resting_AAACGGGCACAACGCC   0.12576722</code></pre>
<p>Amongst the information contained here are: the name (or ID) of each cell, the cell type it belongs to, the individual it was isolated from, the number of genes that were detected inside this cell, etc… All of these variables come from our published study and are only kept here as a reference for comparisons.</p>
<p>Let’s add this information as metadata into our Seurat Object:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb25-1" aria-hidden="true" tabindex="-1"></a>t_cells<span class="sc">@</span>meta.data <span class="ot">&lt;-</span> <span class="fu">cbind</span>(t_cells<span class="sc">@</span>meta.data, cell_anns)</span></code></pre></div>
</div>
</div>
<div id="removing-low-quality-cells-and-genes" class="section level2 hasAnchor" number="3.4">
<h2><span class="header-section-number">3.4</span> Removing low quality cells and genes<a href="practical-studying-cd4-t-cells-from-human-blood.html#removing-low-quality-cells-and-genes" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We now have our data in the right format for analysis. However, we are not quite ready for analysing the biology of this dat set yet. Instead, we first need to remove any technical errors or biases which may skew our conclusions.</p>
<div id="cell-filtering" class="section level3 hasAnchor" number="3.4.1">
<h3><span class="header-section-number">3.4.1</span> Cell filtering<a href="practical-studying-cd4-t-cells-from-human-blood.html#cell-filtering" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>There are many sources of bias and technical error in a single cell experiment, but two of them stand out:
- Apoptotic/dying cells (which could skew our observatoins)
- Multiplets: instances where two or three cells stuck together during the experiment, so that they now seem to be a single cell but really they are two or more cells.</p>
<p>We can remove some of these biases by performing quality filtering.</p>
<p>We will being by removing potentially dying/apoptotic cells. When cells have lysed or are dying, their cell membrane is compromised and pores start appearing along its surface. This causes mRNA molecules to leak outside the cell. However, the mRNAs contained in the mitochondria will not leak as easily, because they have to traverse two sets of membranes: the mitochondrial and the cell membrane. Thus they tend to be retained inside the cell more easily, even when a cell has died.</p>
<p>We can measure this phenomenon by calculating the “mitochondrial proportion”. That is, by estimating what percentage of the molecules detected in a cell come from mitochondrial genes (which are physically located inside the mitochondria). The Seurat package has a function which does this for us automatically:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb26-1" aria-hidden="true" tabindex="-1"></a>t_cells[[<span class="st">&quot;percent.mt&quot;</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(t_cells, <span class="at">pattern =</span> <span class="st">&quot;^MT-&quot;</span>)</span></code></pre></div>
<p>The pattern “^MT-” which we specified tells R that any genes whose name starts with <em>MT-</em> should be counted as mitochondrial genes.</p>
<p>Now that we have estimated this metric. Let’s visualise it. The following code generates a graph where we can see how many genes were detected as expressed in each cell (X axis), as well as the proportion of molecules which come from the mitochondria (Y axis). The lighter the colour, the more cells are located in that area of the graph, and vice versa.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(t_cells<span class="sc">@</span>meta.data, <span class="fu">aes</span>(<span class="at">x=</span>nFeature_RNA, <span class="at">y=</span>percent.mt)) <span class="sc">+</span></span>
<span id="cb27-2"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb27-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hex</span>(<span class="at">bins=</span><span class="dv">70</span>) <span class="sc">+</span></span>
<span id="cb27-3"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">5</span>, <span class="at">linetype=</span><span class="st">&quot;dotted&quot;</span>) <span class="sc">+</span></span>
<span id="cb27-4"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">1600</span>, <span class="at">linetype=</span><span class="st">&quot;dotted&quot;</span>) <span class="sc">+</span></span>
<span id="cb27-5"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb27-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Number of genes per cell&quot;</span>) <span class="sc">+</span></span>
<span id="cb27-6"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb27-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Proportion of reads in mitchondrial genes (%)&quot;</span>) <span class="sc">+</span></span>
<span id="cb27-7"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb27-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb27-8"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb27-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span></code></pre></div>
<p><img src="Introduction-to-single-cell-RNA-seq-data-analysis_files/figure-html/plot_qc_thresholds-1.png" width="672" /></p>
<p>You’ll notice that the majority of cells have around 1500 genes detected, and only about 2.5% of molecules coming from the mitochondria. However, there are some exceptions:</p>
<ul>
<li>Cells above the horizontal line have too many mitochondrial mRNAs. This means that they may be apoptotic</li>
<li>Cells to the right of the vertical line have too many genes detected. This means they may be multiplets (i.e. two cells stuch together rather than one)</li>
</ul>
<p>Thus, we use the following line to remove any cells that have either more than 5% of mitochondrial molecules (likely apoptotic) or more than 1600 genes detected (likely multiplets):</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb28-1" aria-hidden="true" tabindex="-1"></a>t_cells <span class="ot">&lt;-</span> <span class="fu">subset</span>(t_cells, <span class="at">subset =</span> nFeature_RNA <span class="sc">&lt;=</span> <span class="dv">1600</span> <span class="sc">&amp;</span> percent.mt <span class="sc">&lt;=</span> <span class="dv">5</span>)</span></code></pre></div>
<p>This should significantly clean our data set and remove potential confounding factors.</p>
</div>
<div id="gene-filtering" class="section level3 hasAnchor" number="3.4.2">
<h3><span class="header-section-number">3.4.2</span> Gene filtering<a href="practical-studying-cd4-t-cells-from-human-blood.html#gene-filtering" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We now turn our attention to the genes. We have over 14,000 genes, but it’s likely that many of them are present at very low levels or even not present at all.</p>
<p>The following line takes each gene in our matrix and counts in how many cells it is present (i.e. it is not zero).</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb29-1" aria-hidden="true" tabindex="-1"></a>cells_expressing_gene <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">GetAssayData</span>(t_cells) <span class="sc">&gt;</span> <span class="dv">0</span>)</span></code></pre></div>
<p>If we visualise these counts, we’ll notice that a lot of genes are not detected in any cell at all. These are simply taking up unnecessary space in our computer:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(cells_expressing_gene, <span class="at">breaks =</span> <span class="dv">150</span>, <span class="at">xlab=</span><span class="st">&quot;Number of cells expressing the gene&quot;</span>, <span class="at">main=</span><span class="st">&quot;Gene expression distribution&quot;</span>)</span></code></pre></div>
<p><img src="Introduction-to-single-cell-RNA-seq-data-analysis_files/figure-html/plot_cells_expressing_gene-1.png" width="672" /></p>
<p>The following line removes any genes that are simply not found in any cell in the study:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb31-1" aria-hidden="true" tabindex="-1"></a>t_cells <span class="ot">&lt;-</span> <span class="fu">subset</span>(t_cells, <span class="at">features =</span> <span class="fu">rownames</span>(t_cells)[cells_expressing_gene <span class="sc">&gt;</span> <span class="dv">0</span>])</span></code></pre></div>
<p>After filtering out both low quality cells and low quality genes, we are left with 4,979 cells and 14,832 genes:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb32-1" aria-hidden="true" tabindex="-1"></a>t_cells</span></code></pre></div>
<pre><code>## An object of class Seurat 
## 14832 features across 4979 samples within 1 assay 
## Active assay: RNA (14832 features, 0 variable features)</code></pre>
</div>
</div>
<div id="analysing-gene-expression-data" class="section level2 hasAnchor" number="3.5">
<h2><span class="header-section-number">3.5</span> Analysing gene expression data<a href="practical-studying-cd4-t-cells-from-human-blood.html#analysing-gene-expression-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Having done a stringent quality check, we are now ready to start analysing our data.</p>
<p>We being by performing normalisation and data transformation steps. These have the following aims:
- Normalisation: to correct for gross differences in the depth of sequencing between samples. For example, maybe cells sequenced in the first experiment have a lot more molecules than cells sequenced in the second experiment (due to differences in how the sequencer ran). This difference would need to be removed.
- Log-transformation: While some genes have 1 or 2 molecules, others will have 100s of molecules detected. This is a very wide range, and it makes our data have a very skewed distribution. Thus, we apply a logarithm operation, which makes the data nicely centered around a mean value (similarly to a normal distribution).</p>
<p>Both of these operations are performed by the following Seurat function:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb34-1" aria-hidden="true" tabindex="-1"></a>t_cells <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(t_cells)</span></code></pre></div>
<p>Next, we identify which genes tend to vary the most from one cell to another. These tend to be the most interesting genes. If a gene is present at exactly the same level in every cell, then it’s not very interesting. If, instead, it is present at very high levels in one type of cell but at minimnal levels in another one, it could be a very interesting gene with a specific function in the cell type where it is present.</p>
<p>The following function ranks genes according to how much they tend to vary and identifies the top variable (often termed <em>highly variable genes</em> of <em>HVGs</em>. In this case we will get 2,000 of them).</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb35-1" aria-hidden="true" tabindex="-1"></a>t_cells <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(t_cells, <span class="at">selection.method =</span> <span class="st">&quot;vst&quot;</span>, <span class="at">nfeatures =</span> <span class="dv">2000</span>)</span></code></pre></div>
<p>Out of curiosity, what are these genes? The following code generates a simple visualisation telling us how high the level of each gene is on average (X axis) and how much it tends to vary between cells (Y axis).</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify the 10 most highly variable genes</span></span>
<span id="cb36-2"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb36-2" aria-hidden="true" tabindex="-1"></a>top10 <span class="ot">&lt;-</span> <span class="fu">head</span>(<span class="fu">VariableFeatures</span>(t_cells), <span class="dv">10</span>)</span>
<span id="cb36-3"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb36-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-4"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb36-4" aria-hidden="true" tabindex="-1"></a><span class="co"># plot variable features with and without labels</span></span>
<span id="cb36-5"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb36-5" aria-hidden="true" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span> <span class="fu">VariableFeaturePlot</span>(t_cells)</span>
<span id="cb36-6"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb36-6" aria-hidden="true" tabindex="-1"></a>plot2 <span class="ot">&lt;-</span> <span class="fu">LabelPoints</span>(<span class="at">plot =</span> plot1, <span class="at">points =</span> top10, <span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb36-7"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb36-7" aria-hidden="true" tabindex="-1"></a>plot2</span></code></pre></div>
<p><img src="Introduction-to-single-cell-RNA-seq-data-analysis_files/figure-html/find_HVGs-1.png" width="672" /></p>
<p>The most variable genes here are chemokines (CCL5), granulysins (GNLY) and other genes known to distinguish very specific types of T cells. This is encouraging: we are on the right path.</p>
<p>Finally, we scale the data in our expression matrix. This will make the rest of the analysis easier.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb37-1" aria-hidden="true" tabindex="-1"></a>t_cells <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(t_cells, <span class="at">features =</span> <span class="fu">rownames</span>(t_cells))</span></code></pre></div>
</div>
<div id="data-visualisation" class="section level2 hasAnchor" number="3.6">
<h2><span class="header-section-number">3.6</span> Data visualisation<a href="practical-studying-cd4-t-cells-from-human-blood.html#data-visualisation" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Our pre-processing is now over! Let’s explore some interesting biology.</p>
<p>At the moment we have thousands of variables measured across thousands of cells. How do we even begin to make sense of so much information?</p>
<p>One common approach is <strong>dimensionality reduction</strong>, which is based on discarding the majority of these thousands of measurements and instead looking at a combination of those genes which are most informative.</p>
<p>Dimensionality reduction is a complex field of mathematical research, and thus a detailed explanation falls outside of the scope of this course. However, its aim is simple: summarising those thousands of genes (called here <em>dimensions</em>) into a handful of useful variable.</p>
<p>To do this, we use a combination of linear and non-linear techniques.</p>
<div id="principal-component-analysis-pca" class="section level3 hasAnchor" number="3.6.1">
<h3><span class="header-section-number">3.6.1</span> Principal component analysis (PCA)<a href="practical-studying-cd4-t-cells-from-human-blood.html#principal-component-analysis-pca" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We begin with a linear technique called principal component analysis (PCA). This technique is based on linear algebra and rotations around our feature space which identify a small set of gene combinations that summarise most of the information contained in our data set. These gene combinations are called “principal components” or PCs.</p>
<p>The following line performs PCA on our single-cell expression data:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb38-1" aria-hidden="true" tabindex="-1"></a>t_cells <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(t_cells, <span class="at">features =</span> <span class="fu">VariableFeatures</span>(<span class="at">object =</span> t_cells))</span></code></pre></div>
<p>The line above generated a list of principal components. But how many of these are useful? Let’s assess how much information each of them contains:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ElbowPlot</span>(t_cells)</span></code></pre></div>
<p><img src="Introduction-to-single-cell-RNA-seq-data-analysis_files/figure-html/create_elbow_plot-1.png" width="672" /></p>
<p>Note how the variable in the Y axis, which reflects the amount of new information contained by each component (X axis), slowly gets smaller and smaller. In fact, anything beyond the 10-th component is pretty much negligible. So 10 components are enough to summarise the majority of the information in our data set.</p>
<p>We have gone from 1000s of genes to 10 components.</p>
</div>
<div id="component-embedding-with-umap" class="section level3 hasAnchor" number="3.6.2">
<h3><span class="header-section-number">3.6.2</span> Component embedding with UMAP<a href="practical-studying-cd4-t-cells-from-human-blood.html#component-embedding-with-umap" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>While 10 variables is better than 1000s of them, it is still a lot. Thus, we will take those 10 components and “embed” them into a 2-dimensional plot. We do this by using a technique called UMAP, which relies on bending our features space and identifying manifolds within it that could better explain our data structure.</p>
<p>This may sound quite abstract, but the results are very easily intepretable and informative. Let’s have a look.</p>
<p>The code block below takes the first 10 components we just calculated and embeds them with UMAP.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb40-2"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb40-2" aria-hidden="true" tabindex="-1"></a>t_cells <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(t_cells, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span></code></pre></div>
<p>But what does this even mean? Well, let’s see for ourselves what the result looks like:</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(t_cells, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>)</span></code></pre></div>
<p><img src="Introduction-to-single-cell-RNA-seq-data-analysis_files/figure-html/visualise_UMAP_by_gross_cell_type-1.png" width="672" /></p>
<p>The graph above is the result from our PCA + UMAP attempt to reduce dimensions. In this graph, each dot represents a cell. The position of each cell in this plot is determined by a complex combination of the thousands of variables we measured. This can be interperted as follows:</p>
<ul>
<li>If two dots are close to each other, then it means that the two cells have very similar gene expression patterns</li>
<li>If two dots are very far away, then that means the two cells are entirely different frome each other</li>
</ul>
<p>In the plot above, red and blue indicates which cells are memory T cells and which are naive T cells.</p>
<p>In our published study, we did a lot of analysis to understand which T cells belonged to different cell subsets, and that information is included in our metadata table. Thus, let’s now re-do the graph above but colouring cells by their cell type identity:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(t_cells, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">group.by =</span> <span class="st">&quot;cluster.id&quot;</span>)</span></code></pre></div>
<p><img src="Introduction-to-single-cell-RNA-seq-data-analysis_files/figure-html/visualise_UMAP_by_cell_type-1.png" width="672" /></p>
<p>Note how naive cells are on one side of the plot, while memory cells are on the other. Within memory cells, central and effector memory cells separate from each other, and regulatory T cells (which have a completely different cell function and biology) also segregate into their own smaller group.</p>
<p>Seurat also allows us to visualise how active each of the thousands of genes in our matrix is in each cell. The code block below does this for four example genes:
- SELL: Selectin, known to be expressed by naive T cells
- KLRB1: Known to be expressed by effector memory T cells
- PRF1: Marker of T cells with cytotoxic capacity
- IL1R2A: Molecule needed by regulatory T cells to survive</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(t_cells, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">features =</span> <span class="st">&quot;SELL&quot;</span>)</span></code></pre></div>
<p><img src="Introduction-to-single-cell-RNA-seq-data-analysis_files/figure-html/visualise_gene_expression_on_UMAP-1.png" width="672" /></p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(t_cells, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">features =</span> <span class="st">&quot;KLRB1&quot;</span>)</span></code></pre></div>
<p><img src="Introduction-to-single-cell-RNA-seq-data-analysis_files/figure-html/visualise_gene_expression_on_UMAP-2.png" width="672" /></p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(t_cells, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">features =</span> <span class="st">&quot;PRF1&quot;</span>)</span></code></pre></div>
<p><img src="Introduction-to-single-cell-RNA-seq-data-analysis_files/figure-html/visualise_gene_expression_on_UMAP-3.png" width="672" /></p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(t_cells, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">features =</span> <span class="st">&quot;IL2RA&quot;</span>)</span></code></pre></div>
<p><img src="Introduction-to-single-cell-RNA-seq-data-analysis_files/figure-html/visualise_gene_expression_on_UMAP-4.png" width="672" /></p>
<p>Indeed, each of these genes is detected in the cell type where we would expect it to be.</p>
</div>
</div>
<div id="clustering-cells-into-groups" class="section level2 hasAnchor" number="3.7">
<h2><span class="header-section-number">3.7</span> Clustering cells into groups<a href="practical-studying-cd4-t-cells-from-human-blood.html#clustering-cells-into-groups" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>For this example, I provided you with pre-existing cell annotations which tell us which cell belongs to which group. But what if you don’t know that? That’s what happens in most real life applications where we use single-cell techniques.</p>
<p>If we don’t know how many cell types or subtypes we are expecting or how they look like, we can proceed to do a so-called “unsupervised” grouping or clustering of cells. This means that, rather than starting with a pre-conceived assumption about what our cells should look like, we simply compare them to one another and identify patterns. If a group of cells looks remarkably similar to each other, then we group them together.</p>
<p>The Seurat library provides a few functions that we can use to do this. These functions create a “graph” by connecting together cells which are similar to each other, and then identify groups or communities of similar cells, since they are very interconnected in the graph.</p>
<p>The following lines perform unsupervised cell clustering at a high level resolution:</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb47-2"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb47-2" aria-hidden="true" tabindex="-1"></a>t_cells <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(t_cells, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span>
<span id="cb47-3"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb47-3" aria-hidden="true" tabindex="-1"></a>t_cells <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(t_cells, <span class="at">resolution =</span> <span class="fl">0.3</span>)</span></code></pre></div>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 4979
## Number of edges: 157220
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.8535
## Number of communities: 5
## Elapsed time: 0 seconds</code></pre>
<p>Let’s now look at the results from our unsupervised cell grouping:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(t_cells, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>)</span></code></pre></div>
<p><img src="Introduction-to-single-cell-RNA-seq-data-analysis_files/figure-html/visualise_UMAP_by_cluster-1.png" width="672" /></p>
<p>Perhaps unsurprisingly, the groups identified using these algorithm correspond almost exactly with the annotations I provided you with.</p>
</div>
<div id="identifying-cell-type-markers" class="section level2 hasAnchor" number="3.8">
<h2><span class="header-section-number">3.8</span> Identifying cell type markers<a href="practical-studying-cd4-t-cells-from-human-blood.html#identifying-cell-type-markers" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Now that we have grouped cells by similarity, how can we learn more about their biology and understand what each group means?</p>
<p>To do so, we take each gene in our expression table and compare its level between different cell groups: does it tend to be expressed at a significantly higher or lower level in one cell group compared to the remaining ones?</p>
<p>That is what the following function tests (note that this may take a little while to run. After all, this function is doing thousands of tests):</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb50-1" aria-hidden="true" tabindex="-1"></a>t_cell_markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(t_cells, <span class="at">only.pos =</span> <span class="cn">TRUE</span>, <span class="at">min.pct =</span> <span class="fl">0.25</span>, <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span></code></pre></div>
<p>The result of running the line of code above is a list of all genes which are significantly higher or lower in one cell group compared to the others. This looks as follows:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb51-1" aria-hidden="true" tabindex="-1"></a>t_cell_markers <span class="sc">%&gt;%</span></span>
<span id="cb51-2"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb51-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb51-3"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb51-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">slice_max</span>(<span class="at">n =</span> <span class="dv">2</span>, <span class="at">order_by =</span> avg_log2FC)</span></code></pre></div>
<pre><code>## # A tibble: 10 × 7
## # Groups:   cluster [5]
##        p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene  
##        &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt; 
##  1 3.89e-155      0.941 0.375 0.071 5.76e-151 0       LRRN3 
##  2 2.17e-119      0.804 0.64  0.333 3.22e-115 0       CCR7  
##  3 4.91e- 72      0.844 0.499 0.268 7.28e- 68 1       PASK  
##  4 8.68e-125      0.813 0.531 0.197 1.29e-120 1       ITGB1 
##  5 1.04e-274      2.62  0.721 0.118 1.54e-270 2       CCL5  
##  6 1.84e-184      1.57  0.633 0.14  2.73e-180 2       LYAR  
##  7 2.86e-113      2.21  0.689 0.152 4.25e-109 3       LGALS1
##  8 1.10e- 83      1.91  0.956 0.602 1.63e- 79 3       S100A4
##  9 0              3.77  0.833 0.007 0         4       NKG7  
## 10 0              3.54  0.63  0.007 0         4       GNLY</code></pre>
<p>Note how R has calculated the average expression of that gene, as well as performed statistical testing to check if the gene is higher in any one group of cells. This is what the p values represent.</p>
<p>We can understand this information better by looking at it visually using a technique called a <em>heatmap</em>:</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb53-1" aria-hidden="true" tabindex="-1"></a>top_markers <span class="ot">&lt;-</span> t_cell_markers <span class="sc">%&gt;%</span></span>
<span id="cb53-2"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb53-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb53-3"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb53-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">top_n</span>(<span class="at">n =</span> <span class="dv">10</span>, <span class="at">wt =</span> avg_log2FC)</span>
<span id="cb53-4"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb53-4" aria-hidden="true" tabindex="-1"></a><span class="fu">DoHeatmap</span>(t_cells, <span class="at">features =</span> top_markers<span class="sc">$</span>gene) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code></pre></div>
<p><img src="Introduction-to-single-cell-RNA-seq-data-analysis_files/figure-html/visualise_top_cluster_markers-1.png" width="672" /></p>
<p>In this graph, each row is a gene and each (very thin) column is a cell. Black means the gene is not detected in that cell, and bright yellow means it is detected at a high level. Cells have been automatically grouped by similarity and divided into the four populations we identified using unsupervised clustering. Note how some populations (for example cluster 4) have a very clear signature of genes (in this case, genes which are all involved in cytotoxic capacity).</p>
</div>
<div id="conclusion" class="section level2 hasAnchor" number="3.9">
<h2><span class="header-section-number">3.9</span> Conclusion<a href="practical-studying-cd4-t-cells-from-human-blood.html#conclusion" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this course, we used a medium-sized data set of human CD4+ T cells to understand some of the fundamental concepts in single-cell RNA-seq data analysis. We also gained hands on experience on how to analyse this data, including pre-processing, quality checks, visualisation, and unsupervised grouping of cells by similarity.</p>
<p>I hope you found this course useful.</p>
</div>
<div id="r-session" class="section level2 hasAnchor" number="3.10">
<h2><span class="header-section-number">3.10</span> R session<a href="practical-studying-cd4-t-cells-from-human-blood.html#r-session" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="practical-studying-cd4-t-cells-from-human-blood.html#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.1.0 (2021-05-18)
## Platform: aarch64-apple-darwin20 (64-bit)
## Running under: macOS Big Sur 11.5.2
## 
## Matrix products: default
## LAPACK: /Library/Frameworks/R.framework/Versions/4.1-arm64/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices datasets  utils     methods   base     
## 
## other attached packages:
##  [1] Matrix_1.5-4.1     hexbin_1.28.2      SeuratObject_4.1.3 Seurat_4.3.0       forcats_0.5.1      stringr_1.5.0      dplyr_1.1.2       
##  [8] purrr_1.0.1        readr_2.1.2        tidyr_1.3.0        tibble_3.2.1       ggplot2_3.4.2      tidyverse_1.3.2   
## 
## loaded via a namespace (and not attached):
##   [1] readxl_1.3.1           backports_1.4.1        plyr_1.8.8             igraph_1.4.3           lazyeval_0.2.2         sp_1.6-0              
##   [7] splines_4.1.0          listenv_0.9.0          scattermore_1.1        digest_0.6.31          htmltools_0.5.5        fansi_1.0.4           
##  [13] magrittr_2.0.3         tensor_1.5             googlesheets4_1.0.0    cluster_2.1.4          ROCR_1.0-11            tzdb_0.2.0            
##  [19] globals_0.16.2         modelr_0.1.8           matrixStats_0.63.0     spatstat.sparse_3.0-1  colorspace_2.1-0       rvest_1.0.2           
##  [25] ggrepel_0.9.3          haven_2.4.3            xfun_0.39              crayon_1.5.2           jsonlite_1.8.4         progressr_0.13.0      
##  [31] spatstat.data_3.0-1    survival_3.5-5         zoo_1.8-12             glue_1.6.2             polyclip_1.10-4        gtable_0.3.3          
##  [37] gargle_1.2.0           leiden_0.4.3           future.apply_1.11.0    abind_1.4-5            scales_1.2.1           DBI_1.1.2             
##  [43] spatstat.random_3.1-5  miniUI_0.1.1.1         Rcpp_1.0.10            viridisLite_0.4.2      xtable_1.8-4           reticulate_1.28       
##  [49] htmlwidgets_1.6.2      httr_1.4.6             RColorBrewer_1.1-3     ellipsis_0.3.2         ica_1.0-3              pkgconfig_2.0.3       
##  [55] farver_2.1.1           sass_0.4.6             uwot_0.1.14            dbplyr_2.1.1           deldir_1.0-9           utf8_1.2.3            
##  [61] tidyselect_1.2.0       labeling_0.4.2         rlang_1.1.1            reshape2_1.4.4         later_1.3.1            cachem_1.0.8          
##  [67] munsell_0.5.0          cellranger_1.1.0       tools_4.1.0            cli_3.6.1              generics_0.1.3         broom_0.7.12          
##  [73] ggridges_0.5.4         evaluate_0.21          fastmap_1.1.1          yaml_2.3.7             goftest_1.2-3          knitr_1.43            
##  [79] fs_1.6.2               fitdistrplus_1.1-11    RANN_2.6.1             pbapply_1.7-0          future_1.32.0          nlme_3.1-162          
##  [85] mime_0.12              xml2_1.3.3             compiler_4.1.0         rstudioapi_0.13        plotly_4.10.1          png_0.1-8             
##  [91] spatstat.utils_3.0-3   reprex_2.0.1           bslib_0.4.2            stringi_1.7.12         highr_0.10             lattice_0.21-8        
##  [97] vctrs_0.6.2            pillar_1.9.0           lifecycle_1.0.3        jquerylib_0.1.4        spatstat.geom_3.2-1    lmtest_0.9-40         
## [103] RcppAnnoy_0.0.20       data.table_1.14.8      cowplot_1.1.1          irlba_2.3.5.1          httpuv_1.6.11          patchwork_1.1.2       
## [109] R6_2.5.1               bookdown_0.34          promises_1.2.0.1       renv_0.17.3-74         KernSmooth_2.23-21     gridExtra_2.3         
## [115] parallelly_1.36.0      codetools_0.2-19       MASS_7.3-60            assertthat_0.2.1       withr_2.5.0            sctransform_0.3.5     
## [121] parallel_4.1.0         hms_1.1.1              grid_4.1.0             rmarkdown_2.21         googledrive_2.0.0      Rtsne_0.16            
## [127] spatstat.explore_3.2-1 shiny_1.7.4            lubridate_1.8.0</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="getting-started.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
